network/src/testutils/builder.rs

```diff
@@ -0,0 +1,142 @@
+    }
+
+    /// A safety measure to make sure all owners are on purpose
+    pub fn add_owners(mut self, count: u32) -> Self {
+        self.owners += count;
+        self
+    }
+
+    pub fn add_validator(mut self, owner: u32) -> Self {
+        let config = NodeConfig::random_with_template(
+            owner,
+            &NodeConfig::default_for_validator(),
+            &mut self.rng,
+        );
+        let peer_id = config
+            .validator_network
+            .as_ref()
+            .expect("Validator must have a validator network")
+            .peer_id();
+        self.add_node(
+            owner,
+            NodeType::Validator,
+            config,
+            &[
+                PeerNetworkId::new(NetworkId::Validator, peer_id),
+                PeerNetworkId::new(NetworkId::Vfn, peer_id),
+            ],
+        )
+    }
+
+    pub fn add_vfn(mut self, owner: u32) -> Self {
+        let config = NodeConfig::random_with_template(
+            owner,
+            &NodeConfig::default_for_validator_full_node(),
+            &mut self.rng,
+        );
+        let peer_id = config
+            .full_node_networks
+            .iter()
+            .find(|network| network.network_id == NetworkId::Public)
+            .expect("Vfn must have a public network")
+            .peer_id();
+        self.add_node(
+            owner,
+            NodeType::ValidatorFullNode,
+            config,
+            &[
+                PeerNetworkId::new(NetworkId::Vfn, peer_id),
+                PeerNetworkId::new(NetworkId::Public, peer_id),
+            ],
+        )
+    }
+
+    pub fn add_pfn(mut self, owner: u32) -> Self {
+        let config = NodeConfig::random_with_template(
+            owner,
+            &NodeConfig::default_for_public_full_node(),
+            &mut self.rng,
+        );
+        let peer_id = config
+            .full_node_networks
+            .iter()
+            .find(|network| network.network_id == NetworkId::Public)
+            .expect("Pfn must have a public network")
+            .peer_id();
+        self.add_node(
+            owner,
+            NodeType::PublicFullNode,
+            config,
+            &[PeerNetworkId::new(NetworkId::Public, peer_id)],
+        )
+    }
+
+    /// Add a node to the network, ensuring that it doesn't already exist
+    fn add_node(
+        mut self,
+        owner: u32,
+        node_type: NodeType,
+        config: NodeConfig,
+        peer_network_ids: &[PeerNetworkId],
+    ) -> Self {
+        assert!(owner < self.owners);
+
+        let node_id = NodeId { owner, node_type };
+        assert!(!self.nodes.contains_key(&node_id));
+
+        let mut node = Framework::build_node(node_id, config, peer_network_ids);
+        // Add node's sender to every possible node to be connected to
+        // TODO: make this more efficient
+        for (_, other_node) in self.nodes.iter_mut() {
+            if let Some(network_id) = other_node.find_common_network(&node) {
```

Comment:
Should we panic if we can't connect to anything? i.e., no common network?

Reply:
So, this is maybe a little confusing.  Originally, it was top down, so I didn't have to pass all these handles to all the other nodes.

But, now since they're across separate tasks, it's giving each node it's copy of senders for every other node that it can be connected to.  This doesn't mean that it is actually connected.

Panicking causes some other overhead where we have to be very specific about the order we add nodes e.g. we could say everything after the first one needs a handle.

Another option is we panic if we build at the end and a node isn't connected to another.  But then have to do a bunch of logic around if we purposely have missing spaces (for the single node tests).

I went with a simpler approach, when we get around to connecting the nodes, and there are no handles for it, panic (which would tell us we set it up wrong)